openhands/controller/agent_controller.py

```diff
@@ -876,9 +893,16 @@ def _apply_conversation_window(self, events: list[Event]) -> list[Event]:
         if first_user_msg and first_user_msg not in kept_events:
             kept_events = [first_user_msg] + kept_events
 
-        # start_id points to first user message
-        if first_user_msg:
-            self.state.start_id = first_user_msg.id
+        # Verify truncated history fits in context window
+        while (
+            self.context_window is not None
+            and self.agent.llm.get_token_count(kept_events) > self.context_window
+        ):
+            # Need to truncate more - remove oldest non-first-message event
+            if len(kept_events) > 2:  # Keep at least first message and one more event
+                kept_events.pop(1)
```

Comment:
I think we need this to be 2+ events going away: events are usually in order, action, observation, action, observation, and can have message actions between them. The action and obs are linked and they should remain both, or go away both.

Pretty sure this is particularly needed now that we have function calling, where the observation (=tool result) directly references the action (tool call) and LLM APIs error out if not.